src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -622,80 +631,125 @@ private void ProcessGoAwayFrame(FrameHeader frameHeader)
+                throw;
+            }
         }
 
-        private async ValueTask SendSettingsAckAsync()
+        // This method handles flushing bytes to the wire. Writes here need to be atomic, so as to avoid
+        // killing the whole connection. Callers must hold the write lock, but can specify whether or not
+        // they want to release it.
+        private async ValueTask FinishWriteAsync(bool mustFlush, CancellationToken cancellationToken = default, bool releaseWriteLock = true)
         {
-            await AcquireWriteLockAsync().ConfigureAwait(false);
             try
             {
-                _outgoingBuffer.EnsureAvailableSpace(FrameHeader.Size);
-                WriteFrameHeader(new FrameHeader(0, FrameType.Settings, FrameFlags.Ack, 0));
+                // We must flush if the caller requires it, or if there are no other pending writes.
+                if (mustFlush || (releaseWriteLock && _pendingWriters == 0))
+                {
+                    Debug.Assert(_inProgressWrite == null);
+
+                    _inProgressWrite = FlushOutgoingBytesAsync();
+                    
+                    var tcs = new TaskCompletionSourceWithCancellation<bool>();
+                    
+                    if (_inProgressWrite != await Task.WhenAny(_inProgressWrite, tcs.WaitWithCancellationAsync(cancellationToken)).ConfigureAwait(false))
+                    {
+                        throw new OperationCanceledException(cancellationToken);
+                    }
 
-                await FlushOutgoingBytesAsync().ConfigureAwait(false);
+                    await _inProgressWrite.ConfigureAwait(false);
+                }
             }
             finally
             {
-                ReleaseWriteLock();
+                if (releaseWriteLock)
+                {
+                    _writerLock.Release();
+                }
             }
         }
 
-        private async ValueTask SendPingAckAsync(ReadOnlyMemory<byte> pingContent)
+        private async ValueTask AcquireWriteLockAsync(CancellationToken cancellationToken)
         {
-            Debug.Assert(pingContent.Length == FrameHeader.PingLength);
-
-            await AcquireWriteLockAsync().ConfigureAwait(false);
+            Interlocked.Increment(ref _pendingWriters);
             try
             {
-                _outgoingBuffer.EnsureAvailableSpace(FrameHeader.Size + FrameHeader.PingLength);
-                WriteFrameHeader(new FrameHeader(FrameHeader.PingLength, FrameType.Ping, FrameFlags.Ack, 0));
-                pingContent.CopyTo(_outgoingBuffer.AvailableMemory);
-                _outgoingBuffer.Commit(FrameHeader.PingLength);
-
-                await FlushOutgoingBytesAsync().ConfigureAwait(false);
+                await _writerLock.WaitAsync(cancellationToken).ConfigureAwait(false);
             }
             finally
             {
-                ReleaseWriteLock();
+                Interlocked.Decrement(ref _pendingWriters);
+            }
```

Comment:
Could we write this instead as:
```C#
Task acquireLockTask = _writerLock.WaitAsync(cancellationToken);
if (!acquireLockTask.IsCompletedSuccessfully)
{
    Interlocked.Increment(ref _pendingWriters);
    try { await acquireLockTask.ConfigureAwait(false); }
    finally { Interlocked.Decrement(ref _pendingWriters);
}
```
such that we only pay the additional interlocked costs if we didn't successfully acquire the lock already?

Reply:
I've reworked the code to follow this pattern.